File: frontend/src/services/certificate.ts
import apiClient from './api'

export interface Certificate {
  id: number
  common_name: string
  organization?: string
  created_at: string
  expires_at: string
  status: 'active' | 'expired' | 'revoked'
  fingerprint: string
  serial_number: string
}

export interface CertificateCreate {
  common_name: string
  organization?: string
  organizational_unit?: string
  locality?: string
  state_province?: string
  country?: string
  email?: string
  valid_days: number
  key_size: number
  san_dns?: string[]
  san_ip?: string[]
}

export const certificateService = {
  list: async (skip: number = 0, limit: number = 100): Promise<Certificate[]> => {
    const response = await apiClient.get('/certificates/', {
      params: { skip, limit },
    })
    return response.data
  },

  get: async (id: number): Promise<Certificate> => {
    const response = await apiClient.get(`/certificates/${id}`)
    return response.data
  },

  create: async (data: CertificateCreate): Promise<Certificate> => {
    const response = await apiClient.post('/certificates/', data)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/certificates/${id}`)
  },

  download: async (id: number): Promise<Blob> => {
    const response = await apiClient.get(`/certificates/${id}/download`, {
      responseType: 'blob',
    })
    return response.data
  },
}

File: backend/src/ssl_manage/schemas/certificate.py
"""Certificate schemas."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, EmailStr, Field


class CertificateBase(BaseModel):
    """Base certificate schema."""
    
    common_name: str = Field(..., description="Common Name (CN) for the certificate")
    organization: Optional[str] = Field(None, description="Organization name")
    organizational_unit: Optional[str] = Field(None, description="Organizational Unit")
    locality: Optional[str] = Field(None, description="City/Locality")
    state_province: Optional[str] = Field(None, description="State or Province")
    country: Optional[str] = Field(None, description="Country code (2 letters)")
    email: Optional[EmailStr] = Field(None, description="Email address")


class CertificateCreate(CertificateBase):
    """Certificate creation schema."""
    
    valid_days: int = Field(365, description="Certificate validity in days", ge=1, le=3650)
    key_size: int = Field(2048, description="RSA key size", ge=2048, le=4096)
    
    # Subject Alternative Names
    san_dns: Optional[list[str]] = Field(None, description="DNS SANs")
    san_ip: Optional[list[str]] = Field(None, description="IP SANs")


class CertificateResponse(CertificateBase):
    """Certificate response schema."""
    
    id: int
    created_at: datetime
    expires_at: datetime
    serial_number: str
    fingerprint: str
    is_ca: bool = False
    status: str  # active, expired, revoked
    
    class Config:
        from_attributes = True


class CertificateListResponse(BaseModel):
    """Certificate list item schema."""
    
    id: int
    common_name: str
    organization: Optional[str]
    created_at: datetime
    expires_at: datetime
    status: str
    
    class Config:
        from_attributes = True


class CertificateDownload(BaseModel):
    """Certificate download response."""
    
    certificate_pem: str
    private_key_pem: str
    certificate_chain_pem: Optional[str] = None